httpbin/middleware.go

```diff
@@ -163,22 +163,19 @@ type Observer func(result Result)
 
 // StdLogObserver creates an Observer that will log each request in structured
 // format using the given stdlib logger
-func StdLogObserver(l *log.Logger) Observer {
+func StdLogObserver(l *slog.Logger) Observer {
 	const (
-		logFmt  = "time=%q status=%d method=%q uri=%q size_bytes=%d duration_ms=%0.02f user_agent=%q client_ip=%s"
 		dateFmt = "2006-01-02T15:04:05.9999"
 	)
 	return func(result Result) {
-		l.Printf(
-			logFmt,
-			time.Now().Format(dateFmt),
-			result.Status,
-			result.Method,
-			result.URI,
-			result.Size,
-			result.Duration.Seconds()*1e3, // https://github.com/golang/go/issues/5491#issuecomment-66079585
-			result.UserAgent,
-			result.ClientIP,
-		)
+		l.With("time", time.Now().Format(dateFmt)).
+			With("status", result.Status).
+			With("method", result.Method).
+			With("uri", result.URI).
+			With("size_bytes", result.Size).
+			With("duration_ms", result.Duration.Seconds()*1e3).
+			With("user_agent", result.UserAgent).
+			With("client_ip", result.ClientIP).
+			Info("request")
```

Comment:
Hey, thanks for the review and the insightful comments. Sorry for the late reply.

Regarding the suggestion to drop the request message:

I understand the point you're making, but it's a bit tricky with the `slog`. Unfortunately, with slog, there's no built-in way to log structured data without including a msg field when using the default handlers. I could set it as an empty string, but that looks awkward and doesn't seem like a good solution. Here's an example of what that would look like:

```
time=2024-08-26T00:54:24.005+02:00 level=INFO msg="" time=2024-08-26T00:54:24.0053 status=200 method=GET uri=/ size_bytes=11839 duration_ms=0.083083 user_agent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/127.0.0.0 Safari/537.36" client_ip=127.0.0.1:63536
```

If I understood correctly, you're suggesting parsing all the fields within the msg field for better readability, especially for tools like DataDog. However, in my experience with Elasticsearch, it's generally more effective and considered best practice to keep each field separate. This approach enhances searchability and performance.

Do you think there's room for a compromise, or should we stick with the current approach given the limitations?

Reply:
> If I understood correctly, you're suggesting parsing all the fields within the msg field for better readability, especially for tools like DataDog. However, in my experience with Elasticsearch, it's generally more effective and considered best practice to keep each field separate. This approach enhances searchability and performance.

Not quite.  What I'm suggesting is that if we are going to include a message, we should make it useful on its own.  What we do at work is duplicate some of the structured fields into a human-friendly traditional web log style message, like so (using [zerolog](https://github.com/rs/zerolog) rather than slog):

```go
// Use DataDog's standard log attributes:
// https://docs.datadoghq.com/logs/processing/attributes_naming_convention/#default-standard-attribute-list
hlog.FromRequest(r).
	WithLevel(lvl).
	Int64("duration", duration.Nanoseconds()).
	Str("http.host", r.Host).
	Str("http.method", r.Method).
	Int("http.status_code", status).
	Str("http.url", r.URL.String()).
	Str("http.useragent", r.Header.Get("User-Agent"))).
	Int64("network.bytes_read", r.ContentLength).
	Int("network.bytes_written", size).
	Str("network.client.ip", getClientIP(r)))).
	Msgf("%d %s %s (%.1fms)", status, r.Method, r.RequestURI, durationMS)
```

So a log message might end up looking like

```
200 GET /example 12ms
```

where the status, method, request URI and duration are still included directly in the structured fields and can be queried/graphed/filtered/etc.

Reply:
Thanks for clarification @mccutchen I was kinda confused about what you were asking. I hope now it should be like you wanted 🙂